Canonicalization scorer

jcollins-g opened the pull request:
Fixes #1370.  This is a new canonicalization scorer to enable dartdoc to make good decisions for where to put docs for a reexported element.   The scorer uses paths much like @Hixie suggested in the fifth bullet point of [this comment](https://github.com/dart-lang/dartdoc/issues/1334#issuecomment-286159726), with some additions (variants on the rest of that comment are already implemented).   It eliminates all but 4 of the ambiguous reexport warnings from angular2, dropping the number of ambiguous reexport warnings by 97%.  The last few can be fixed with a new directive, `@canonicalFor`, that can be used to disambiguate or override bad scorer decisions.

I did manually go through the hundred and thirty three warnings this eliminated and am convinced a human would agree with me as to where they should be canonical -- the last few that display I wasn't completely sure of.  The flag --ambiguous-reexport-scorer-min-confidence will allow you to adjust when these warnings appear if you think the scorer is broken.

Old warnings:
![before](https://cloud.githubusercontent.com/assets/14116827/26746695/05f82f1a-47a6-11e7-92b6-a1562d73f0e4.png)
.... and they scroll off the screen.

New warnings:
![after](https://cloud.githubusercontent.com/assets/14116827/26746699/0d78e608-47a6-11e7-8202-5159d6226189.png)

To support testing the `@canonicalFor` feature I had to make minor changes elsewhere in dartdoc.  Additionally, the warning system has been enhanced to permit enhanced warning descriptions -- the ambiguous reexport warning is particularly challenging to understand and I felt without some more infrastructure to help the user even the smaller number of warnings wouldn't be actionable.

@chalin @kwalrath

kwalrath commented:
Can't wait to see how this improves the Angular docs!

Hixie commented:
oooOOOooo!

Do we need to use `@canonicalFor` anywhere in Flutter? What does it do? :-)

chalin commented:
Very nice! I'm also looking forward to see this in action.

devoncarew commented on bin/dartdoc.dart:
```diff
@@ -152,21 +152,21 @@ main(List<String> arguments) async {
 
   setConfig(
-      addCrossdart: addCrossdart,
+      addCrossdart: args['add-crossdart'] as bool,
```
I'm curious if the analyzer needs this casting - do you get a warning without the `as bool`?

devoncarew commented on bin/dartdoc.dart:
```diff
@@ -266,6 +266,14 @@ ArgParser _createArgsParser() {
           "Generates `index.json` with indentation and newlines. The file is larger, but it's also easier to diff.",
       negatable: false,
       defaultsTo: false);
+  parser.addOption('ambiguous-reexport-scorer-min-confidence',
```
This seems like a super long name for a flag. Is this even a flag that user's should toggle? Should we just decide on a good value for this?

If you need a flag for dev purposes, you can have a hidden option with the `hide: true` param.

jcollins-g replied:
Added `hide: true`.

devoncarew commented on lib/src/model.dart:
```diff
@@ -2759,53 +2998,78 @@ enum PackageWarning {
+      "ambiguous-reexport",
+      "A symbol is exported from private to public in more than one library and dartdoc can not determine which one is canonical",
+      [
+        "Use {@canonicalFor @@name@@} in the desired library's documentation to resolve",
```
`@canonicalDeclarationFor`? `@canonicalDocLocation`? `@docLocationFor`? `@canonicalLocationFor`? `@dartdocCanonical`?

jcollins-g replied:
I picked `@canonicalFor` since the tag is to appear in the library which is canonical for the symbol given as a parameter.  It'll look like this:

```
/// Some docs for foo.
/// {@canonicalFor bar.Baz}
library foo;
```

Since we're not specifying a location as a parameter to the tag I tend to drop out options 2-4.  I'm not sure what the first one would mean in this context.  The last one seems to add unnecessary information (AFAIK there are no similar comment tags for other tools).

jcollins-g commented:
--auto-include-deps was broken and I didn't notice it until the upload (we added src/mylib from the test package).  I have no idea how it managed to work before, it probably was accidental.  So,  added a checker to isPrivate to never declare public something in lib/src of a package.

jcollins-g commented:
@Hixie: The only ambiguous reexport warning in flutter is for Matrix4, since it is reexported from other packages directly.   Flutter is constructed in such a way that it never crosses the private/public export boundary in more than one public library for a given symbol, so the canonicalization code knows to prioritize the library where the boundary was crossed and ignore other reexports.

For example:
lib/reexporter_of_foo.dart <--- lib/foo.dart <--- lib/src/my_private_library.dart
lib/another_reexporter_of_foo.dart <--- lib/foo.dart

As long as your reexports are structured this way, dartdoc knows to prioritize lib/foo.dart for symbols originally from lib/src/my_private_library.dart.  Both the Dart SDK and flutter seem to be built this way.

Angular2 has a lot of things that look like this (simplified):

lib/foo.dart <--- lib/src/my_private_library.dart
lib/library.dart <--- lib/src/my_private_library.dart
lib/another_library.dart <--- lib/src/my_private_library.dart
lib/stuff_and_things.dart <--- lib/src/my_private_library.dart

With this structure, dartdoc used to just give up and pick one of these arbitrarily as canonical, emitting a warning.  The scorer enables it to guess better.  It sets values to things from the library and directory structure, whether a library is deprecated, and so on, to determine where it should be canonical.  But it's possible to guess wrong, and @canonicalFor makes up the difference there.

pq commented on lib/src/model.dart:
```diff
@@ -1299,6 +1309,25 @@ class Library extends ModelElement {
+  /// documented with this library, but these ModelElements and names correspond
+  /// to the defining library where each originally came from with respect
+  /// to inheritance and reexporting.  Most useful for error reporting.
+  Iterable<String> get allOriginalModelElementNames {
```
Curious why you're not returning a `List` here.  Is this to prevent write access?

pq commented on lib/src/model.dart:
```diff
@@ -1705,6 +1819,41 @@ class Method extends ModelElement
+  }
+
+  @override
+  String toString() {
```
Short form? 

    => "${library.name}: ${score.toStringAsPrecision(4)} - ${reasons.join(', ')}";

pq commented on lib/src/model.dart:
```diff
@@ -1865,6 +2014,67 @@ abstract class ModelElement implements Comparable, Nameable, Documentable {
+          .where((s) => s.isNotEmpty));
+  }
+
+  Set<String> get namePieces {
```
Similarly here.  I prefer the `=>` form but I guess it's a matter of taste. 🤔

pq commented on lib/src/model.dart:
```diff
@@ -2024,6 +2255,10 @@ abstract class ModelElement implements Comparable, Nameable, Documentable {
   }
 
+  List<ScoredCandidate> scoreCanonicalCandidates(List<Library> libraries) {
+    return libraries.map((l) => scoreElementWithLibrary(l)).toList()..sort();
```
Ditto here.

pq replied:
But of course, feel free to ignore!

pq commented on lib/src/model.dart:
```diff
@@ -2759,53 +2998,78 @@ enum PackageWarning {
+
+  PackageWarningHelpText(this.warning, this.warningName, this.shortHelp,
+      [this.longHelp]) {
+    if (this.longHelp == null) this.longHelp = [];
```
I think you could do `this.longHelp ??= []`...

pq commented on lib/src/model.dart:
```diff
@@ -2891,29 +3170,43 @@ class PackageWarningCounter {
-    _countedWarnings.putIfAbsent(element, () => new Set());
-    _countedWarnings[element].add(warningData);
-    _writeWarning(kind, fullMessage);
+    _countedWarnings.putIfAbsent(element?.element, () => new Set());
```
It looks like you're allowing for mappings with `null` keys.  Is that right?

pq commented on test/model_test.dart:
```diff
@@ -218,6 +235,40 @@ void main() {
+
+    test('@canonicalFor directive works', () {
+      expect(SomeOtherClass.canonicalLibrary, equals(reexportOneLib));
+      expect(SomeClass.canonicalLibrary, equals(reexportTwoLib));
```
I think you can elide the `equals`?

jcollins-g commented:
Realizing I didn't reply here to @devoncarew about reducing SDK errors. We talked about this briefly offline, but for the record:

This is only for ambiguous reexport warning.  When determining canonicalization for a given symbol, there are three possible paths:

1) There are multiple canonical locations and we have to choose between them.
2) There is exactly one possible canonical location.
3) There is no canonicalization available.

Earlier work in #1390 and #1368 implemented the second item above, and partially implemented the first in a way that worked well for Flutter but not AngularDart.  This PR improves the first case to the maximum extent I can think of for now and gives control to users for disambiguating cases it can't handle.  Improving the third case (no canonicalization found) is still in the queue and will require knowledge of packages not being documented.

devoncarew commented:
My comments mostly just applied to the UI of the tool; lgtm for my part
